src/lib/optimizer/strategy/dependent_group_by_reduction_rule.hpp

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include "abstract_rule.hpp"
+#include "expression/abstract_expression.hpp"
+#include "expression/lqp_column_expression.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+
+// Removes columns from list of group by columns when they are functionally dependent from column(s) of the grouping.
+// This is currently only the case, when both the primary key and other columns of the same table are grouped.
+// This is regularly the case, when the columns are supposed to be later accessed (e.g., see usage of "c_acctbal" in
+// TPC-H query 10). As "c_acctbal" does not need to be in the grouping (since the primary key c_custkey is present),
+// the column in added as `ANY(c_acctbal)` to the aggregation list. ANY() selects "any" value from the list of values
+// per group (since we group by the primary key in this case, the group is ensured to be of size one).
+// This rule implements choke point 1.4 of "TPC-H Analyzed: Hidden Messages and Lessons Learned from an Influential
+// Benchmark" (Boncz et al.). Due to the current lack of foreign key suppport, not all queries that are listed in the
+// paper can be optimized.
+// When the group by reduction changes the column order and this order is also the final column order of the plan, a
+// projection is appended as the new root to restore the column order.
```

Comment:
Not sure if this is understood by others. I made my own attempt, maybe we can merge the two into something good:

> In SQL, the returned columns of an aggregate are either "group by columns" or "aggregate columns, e.g., sum(x)". A query to select the customer's id, their name, and the number of orders could look like this:
SELECT customer.id, customer.name, COUNT(*) FROM customer JOIN orders ON ... GROUP BY customer.id, customer.name
Note how customer.name is included as a group by column, even though the groups are already exhaustively defined by `GROUP BY customer.id`. The reason for that is the functional dependency `id => name`, or, more practically speaking, the fact that `customer.id` is a primary key.
This rule identifies cases where functionally dependent columns are given as group by columns while their primary key column(s) is a/are also group by column(s). It then replaces that column (e.g., customer.name) with a dummy aggregate function ANY(customer.name). This reduces the grouping cost in the aggregate operator.
As the aggregate operators first produce all group by column, followed by all aggregate columns (including ANY), the order might be changed. Unless a following Projection redefines the order anyway, a new Projection is inserted to restore the original column order.

Reply:
Sounds good. Merged your text with mine.